File: Patterns/Design Patterns/Singleton/SingletonCode/C++/Singleton.cpp
#include <iostream>
#include <cstdlib>
#include <string>
using namespace std;

class Singleton {

    private:
        static Singleton* instanceptr;
        string name;

        Singleton() {

        }

    public:

        static Singleton* getInstance() {
            if (instanceptr == NULL) {
                instanceptr = new Singleton();
                return instanceptr;
            }
            else {
                return instanceptr;
            }
        }
        void setvalues(string n) {
            name = n;
        }
        void print() {
            cout << name << endl;
        }
};

Singleton* Singleton::instanceptr = NULL;

int main() {
    Singleton* a = Singleton::getInstance();

    a->setvalues("Alex");
    a->print();

    Singleton* b = Singleton::getInstance();
    b->print();

    cout << "Memory location for a: " << a << endl;
    cout << "Memory location for b: " << b << endl;

    return 0;
}

File: CourseWork/CS 326/Patterns/Design Patterns/Singleton/SingletonCode/Java/Singleton.java
// Alex Miller
// Singleton Pattern Example
import java.io.*;
import java.util.*;

public class Singleton { // main class
	public static void main(String[] args) { // main metho
		SingletonObj a = SingletonObj.getInstance(); // create SingletonObj a
		a.setName("Alex"); // set the name of a to "Alex"
		a.Print(); // print a

		SingletonObj b = SingletonObj.getInstance(); // create SingletonObj b
		b.Print(); // print b, also prints "Alex"

		// both of the memory locations are the same since they are the same object
		System.out.println("Memory location of a: " + a);
		System.out.println("Memory location of b: " + b);
	}
}
class SingletonObj { // Singleton Object class
	private static SingletonObj Inst = null; // initialize to null
	private String name; // data

	public static SingletonObj getInstance() { // get the instance of the Singleton Object
		if (Inst == null) { // if Inst is null then make a new SingletonObj, if not return the one already created
			Inst = new SingletonObj();
			return Inst;
		}
		else return Inst;
	}
	public void setName(String n) { // set the name data of the SingletonObj
		name = n;
	}
	public void Print() { // print the name of the SingletonObj
		System.out.println(name);
	}
}

